src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs

```diff
@@ -0,0 +1,262 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Roslyn.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class DataFlowPass
+    {
+        /// <summary>
+        /// Holds info from the assignment pass done by
+        /// <see cref="LocalFunctionAssignmentPass"/>.
+        /// </summary>
+        private LocalFuncAssignmentResults _localFuncAssignmentResults;
+
+        /// <summary>
+        /// True if we're currently recording reads, rather than reporting
+        /// diagnostics.
+        /// </summary>
+        private bool _recordingReads = false;
+
+        private SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> _readVars =
+            new SmallDictionary<LocalFunctionSymbol, LocalFuncInfo>();
+
+
+        protected class LocalFuncInfo
+        {
+            public bool IsDirty { get; set; }
+            public PooledHashSet<int> UsedVars =
+                PooledHashSet<int>.GetInstance();
+
+            public void Free()
+            {
+                UsedVars.Free();
+            }
+        }
+
+        protected class LocalFuncAssignmentResults
+        {
+            public SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> AssignedVars { get; }
+            public ImmutableArray<VariableIdentifier> VariableSlots { get; }
+
+            public LocalFuncAssignmentResults(
+                SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> assignedVars,
+                ImmutableArray<VariableIdentifier> variableSlots)
+            {
+                AssignedVars = assignedVars;
+                VariableSlots = variableSlots;
+            }
+
+            public void Free()
+            {
+                FreeLocalFuncInfos(AssignedVars.Values);
+            }
+        }
+
+        private static void FreeLocalFuncInfos<T>(T infos)
+            where T : IEnumerable<LocalFuncInfo>
+        {
+            foreach (var info in infos)
+            {
+                info.Free();
+            }
+        }
+
+        protected static bool HasDirtyLocalFunctions<T>(T localFuncInfos)
+            where T : IEnumerable<LocalFuncInfo>
+        {
+            foreach (var info in localFuncInfos)
+            {
+                if (info.IsDirty)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        protected static void ClearDirtyBits<T>(T localFuncInfos)
```

Comment:
>ClearDirtyBits [](start = 30, length = 14)

You might be able to reduce by one pass if you track whether the data was read after it was written (it is only dirty if changed after it was read).